libraries/AP_Beacon/AP_Beacon_Nooploop.cpp

```diff
@@ -0,0 +1,238 @@
+            _state = ParseState::LEN_L;
+        } else {
+            _state = ParseState::HEADER;
+        }
+        break;
+
+    case ParseState::H54_FUNCTION_MARK:
+        if (b == NOOPLOOP_FUNCTION_MARK_SETTING_FRAME0) {
+            _msgbuf[1] = b;
+            _msg_len++;
+            _crc_expected += b;
+            _state = ParseState::SF0_PAYLOAD;
+        } else {
+            _state = ParseState::HEADER;
+        }
+        break;
+
+    case ParseState::LEN_L:
+        _msgbuf[2] = b;
+        _msg_len++;
+        _crc_expected += b;
+        _state = ParseState::LEN_H;
+        break;
+
+    case ParseState::LEN_H:
+        // extract and sanity check frame length
+        _frame_len = (uint16_t)b << 8 | _msgbuf[2];
+        if (_frame_len > NOOPLOOP_NODE_FRAME2_FRAMELEN_MAX) {
+            _state = ParseState::HEADER;
+        } else {
+            _msgbuf[3] = b;
+            _msg_len++;
+            _crc_expected += b;
+            _state = ParseState::NF2_PAYLOAD;
+        }
+        break;
+
+    case ParseState::NF2_PAYLOAD:
+        // add byte to buffer if there is room
+        if (_msg_len < NOOPLOOP_MSG_BUF_MAX) {
+            _msgbuf[_msg_len] = b;
+        }
+        _msg_len++;
+        if (_msg_len >= _frame_len) {
+            _state = ParseState::HEADER;
+            // check crc
+            if (b == _crc_expected) {
+                return MsgType::NODE_FRAME2;
+            } else {
+                return MsgType::INVALID;
+            }
+        } else {
+            _crc_expected += b;
+        }
+        break;
+
+    case ParseState::SF0_PAYLOAD:
+        // add byte to buffer if there is room
+        if (_msg_len < NOOPLOOP_MSG_BUF_MAX) {
+            _msgbuf[_msg_len] = b;
+        }
+        _msg_len++;
+        if (_msg_len >= 128) {
+            _state = ParseState::HEADER;
+            // check crc
+            if (b == _crc_expected) {
+                return MsgType::SETTING_FRAME0;
+            } else {
+                return MsgType::INVALID;
+            }
+        } else {
+            _crc_expected += b;
+        }
+        break;
+    }
+
+    return MsgType::INVALID;
+}
+
+void AP_Beacon_Nooploop::parse_node_frame2()
+{
+    // a message has been received
+    _last_update_ms = AP_HAL::millis();
+
+    // estimated precision for x,y position in meters
+    const float precision_x = _msgbuf[NOOPLOOP_NODE_FRAME2_PRECISION_X] * 0.01;
+    const float precision_y = _msgbuf[NOOPLOOP_NODE_FRAME2_PRECISION_Y] * 0.01;
+    //EKF's estimate goes very bad if the error value sent into the EKF is unrealistically low. ensure it's never less than a reasonable value
+    const float pos_err = MAX(0.1f, sqrtf(sq(precision_x)+sq(precision_y)));
+
+    // x,y,z position in m*1000 in ENU frame
+    const int32_t pos_x = ((int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSX+2] << 24 | (int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSX+1] << 16 | (int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSX] << 8) >> 8;
```

Comment:
This is odd.  Why not shift the parts a little less and avoid the shift at the end?

Reply:
@peterbarker,

I think it may make a difference for negative numbers.

Reply:
Following lines are extracted from user manual

>  According to the LinkTrack product data, in order to represent more data with as few bytes as
possible, the floating point number is represented by shaping and transmitted through the protocol
frame. Therefore, when unpacking, the shaping data with magnification is actually floating-point,
which needs to be divided by the multiplying power of identification in NLink.
In particular, for type int24, we need to first convert it to type int32. To keep the sign the same, we
divide by 256 after move it to the left. For example, for location data, we use int24 to express it, with a
multiplier of 1000. The parsing code is as follows:
uint8_t byte[] = {0xe6,0x0e,0x00};// represents a decimal value:3.814
int32_t temp = (int32_t)(byte[0] << 8 | byte[1] << 16 | byte[2] << 24) / 256;
float result = temp/1000.0f;
The same goes for uint24 type variables.